### PR TITLE
Dynamic vote

### PR DESCRIPTION
**Pull request by achamayou**
Implements #365, still needs more testcases, but ready for early review.

In short, this changes governance so that vote counting happens in the constitution code. This allows having several different kinds of members, and in particular "operators" or "operating members", which can add/remove nodes and update code but are otherwise unable to vote on other proposals.

### CODE REVIEW

**Comment by cimetrics** [comment]
images

**Comment by cimetrics** [comment]
images

**Comment by cimetrics** [comment]
images

**Comment by cimetrics** [comment]
images

**Comment by cimetrics** [comment]
images

File: src/runtime_config/operator_gov.lua [hunk 1]
```diff
@@ -0,0 +1,130 @@
...
+  end
+
+  -- a single operator can pass an operator vote
+  if operator_vote then
+    return operator_votes > 0
+  end
+
+  return false]],
+
+  environment_proposal = [[
```
Comment (olgavrou) [inline src/runtime_config/operator_gov.lua]: Do we need to define these environment proposals every time? Can we register a default one and override if it is defined?

Comment (achamayou) [reply src/runtime_config/operator_gov.lua]: Every time? This is only defined once per constitution. We could register a default one, as could we with the constitution itself, but having all the constitution explicitly in one file also seems good from an audit perspective (users should read it at least once!). Users wouldn't want to pick up a new environment_proposal without refreshing their constitution accordingly (the two may not be compatible!), so if we start splitting them out, we have to worry about versions etc.

This seems to be equivalent to the discussion about whether we should have the simple and operator constitutions separate, or as one, parametrised version. If we split them, we have some duplication, but the simple one really is simple. If we don't, we don't have duplication, but people who go for simple have to audit a bunch of extra code they don't need.

Either way, we have to test both. We're also not heading towards having N samples, and we definitely don't want to worry about dependency management and packaging for Lua in the KV (at least not now!).

So in short, I think I'd rather not. But given compelling arguments, I could change my mind :)

**Comment by cimetrics** [comment]
images

File: src/node/rpc/test/membervoting_test.cpp [hunk 2]
```diff
@@ -1008,6 +1010,308 @@ TEST_CASE("Add user via proposed call")
...
+      get_proposal(rpc_ctx, frontend, proposal_id, 1);
+
+    const auto& votes = proposal.result.votes;
+    CHECK(votes.size() == 3);
+
+    const auto operator_vote = votes.find(operator_id);
+    CHECK(operator_vote != votes.end());
+    CHECK(operator_vote->second == vote_for);
+
+    const auto proposer_vote = votes.find(1);
```
Comment (eddyashton) [inline src/node/rpc/test/membervoting_test.cpp]: Can we use `proposer_id` and `voter_id` rather than `1` and `2`?

File: src/node/rpc/test/membervoting_test.cpp [hunk 3]
```diff
@@ -1008,6 +1010,308 @@ TEST_CASE("Add user via proposed call")
...
+}
+
+TEST_CASE("Passing operator vote")
+{
+  NetworkTables network;
+  GenesisGenerator gen(network);
+  gen.init_values();
+  auto new_kp = tls::make_key_pair();
+  auto new_ca = new_kp->self_sign("CN=new node");
+  NodeInfo ni = {"", "", "", "", new_ca, {}};
```
Comment (eddyashton) [inline src/node/rpc/test/membervoting_test.cpp]: Minor: Would prefer

```
NodeInfo ni;
ni.cert = new_ca;
```

File: src/runtime_config/gov.lua [hunk 4]
```diff
@@ -2,32 +2,48 @@
...
+
+  member_votes = 0
+
+  for member, vote in pairs(votes) do
+    if vote then
+      member_votes = member_votes + 1
+    end
+  end
+
+  -- count active members, excluding operators
```
Comment (jumaffre) [inline src/runtime_config/gov.lua]: This comment only applies to the operator version of the constitution.

File: src/node/rpc/test/membervoting_test.cpp [hunk 5]
```diff
@@ -1008,6 +1010,308 @@ TEST_CASE("Add user via proposed call")
...
+
+    json proposej = create_json_req(Propose::In{proposal, node_id, vote_for}, "propose");
+    ccf::SignedReq sr(proposej);
+
+    Store::Tx tx;
+    enclave::RPCContext rpc_ctx(operator_id, operator_cert);
+    Response<Propose::Out> r =
+      frontend.process_json(rpc_ctx, tx, operator_id, proposej, sr).value();
+
+    CHECK(r.result.completed);
```
Comment (jumaffre) [inline src/node/rpc/test/membervoting_test.cpp]: Is it also worth reading the nodes tables and checking that the node is now `TRUSTED`?

File: src/node/rpc/test/membervoting_test.cpp [hunk 6]
```diff
@@ -1008,6 +1010,308 @@ TEST_CASE("Add user via proposed call")
...
+    const Response<Proposal> proposal =
+      get_proposal(rpc_ctx, frontend, proposal_id, 1);
+
+    const auto& votes = proposal.result.votes;
+    CHECK(votes.size() == 1);
+
+    const auto proposer_vote = votes.find(operator_id);
+    CHECK(proposer_vote != votes.end());
+    // TODO: find out why this doesn't work
+    //CHECK(proposer_vote->second == vote_for);
```
Comment (jumaffre) [inline src/node/rpc/test/membervoting_test.cpp]: What's the current value of `proposer_vote->second`? I'm surprised it doesn't work either.

File: src/node/rpc/test/membervoting_test.cpp (hunk 2 above)
Comment (achamayou) [reply src/node/rpc/test/membervoting_test.cpp]: Sure, but these are member one and two, as opposed to the operator. The important distinction here is members vs operators, not proposer vs voter.

File: src/node/rpc/test/membervoting_test.cpp (hunk 3 above)
Comment (achamayou) [reply src/node/rpc/test/membervoting_test.cpp]: Ok

File: src/runtime_config/gov.lua (hunk 4 above)
Comment (achamayou) [reply src/runtime_config/gov.lua]: Yes!

File: src/node/rpc/test/membervoting_test.cpp (hunk 5 above)
Comment (achamayou) [reply src/node/rpc/test/membervoting_test.cpp]: To test the voting of the new node, yes, to test the operator-specific logic, no.

File: src/node/rpc/test/membervoting_test.cpp (hunk 6 above)
Comment (achamayou) [reply src/node/rpc/test/membervoting_test.cpp]: It does work, there was a bug in the test :)

File: src/runtime_config/operator_gov.lua [hunk 7]
```diff
@@ -0,0 +1,130 @@
...
+        if call.args[sensitive_table] then
+          -- require unanimity of non-operating members
+          return member_votes == members_active 
+        end
+      end
+    end
+  end
+
+  -- a vote is an operator vote if it's only making operator calls
+  operator_vote = true;
```
Comment (jumaffre) [inline src/runtime_config/operator_gov.lua]: Minor: maybe it's just me but I'd rather have `operator_vote = false` here and the condition flipped in the for loop below. Also, no need for `;`.

**Comment by cimetrics** [comment]
images

File: src/runtime_config/operator_gov.lua [hunk 8]
```diff
@@ -0,0 +1,130 @@
...
+  STATE_ACTIVE = 1
+
+  tables["ccf.members"]:foreach(function(member, details) 
+    if details["status"] == STATE_ACTIVE and not is_operator(tostring(member)) then
+      members_active = members_active + 1 
+    end 
+  end)
+
+  -- check for raw_puts to sensitive tables
+  SENSITIVE_TABLES = {"ccf.whitelists", "ccf.gov_scripts"}
```
Comment (jumaffre) [inline src/runtime_config/operator_gov.lua]: We should add some tests for writing on sensitive tables at some point (outside of this PR).

**Comment by jumaffre** [comment]
We should document this briefly in https://microsoft.github.io/CCF/governance.html

**Comment by cimetrics** [comment]
images

File: src/runtime_config/operator_gov.lua (hunk 7 above)
Comment (achamayou) [reply src/runtime_config/operator_gov.lua]: @jumaffre votes are only operator votes if all actions are operator actions, so this seems more efficient and logical, am I missing something?

File: src/runtime_config/operator_gov.lua (hunk 8 above)
Comment (achamayou) [reply src/runtime_config/operator_gov.lua]: We have one at line 788

**Comment by cimetrics** [comment]
images

**Comment by cimetrics** [comment]
images

**Comment by cimetrics** [comment]
images

**Comment by cimetrics** [comment]
images

**Comment by cimetrics** [comment]
images

File: sphinx/source/governance.rst [hunk 9]
```diff
@@ -123,4 +118,54 @@ Once the proposal has been accepted, nodes running the new code are authorised j
...
 
-.. note:: The safest way to restart or replace nodes is by stopping a single node running the old version and starting a node running the new version as a sequence of operations, in order to avoid a situation in which most nodes have been stopped, and new nodes will not be able to join since it would be impossible to reach a majority of nodes agreeing to accept new nodes (this restriction is imposed by the consensus algorithm).
\ No newline at end of file
+.. note:: The safest way to restart or replace nodes is by stopping a single node running the old version and starting a node running the new version as a sequence of operations, in order to avoid a situation in which most nodes have been stopped, and new nodes will not be able to join since it would be impossible to reach a majority of nodes agreeing to accept new nodes (this restriction is imposed by the consensus algorithm).
+
+
+Models
+------
+
+The entities operating a CCF network do not necessarily overlap with the members of that network. Although the scriptability of the governance model effectively allows a large number of possible arrangements, the following two schemes seem most likely:
```
Comment (jumaffre) [inline sphinx/source/governance.rst]: entities operating -> operators?

Comment (jumaffre) [reply sphinx/source/governance.rst]: scriptability -> flexibility?

File: sphinx/source/governance.rst [hunk 10]
```diff
@@ -123,4 +118,54 @@ Once the proposal has been accepted, nodes running the new code are authorised j
...
+
+Non-member operators
+````````````````````
+
+It is possible for a set of operators to host a CCF network without being members. These operators could:
+
+- Start the network
+- Hand it over to the members for them to Open (TR, Section IV B)
```
Comment (jumaffre) [inline sphinx/source/governance.rst]: We should reference other sections of this doc here, rather than the TR.

File: sphinx/source/governance.rst (hunk 9 above)
Comment (achamayou) [reply sphinx/source/governance.rst]: I was deliberately avoiding the word, because I use operator/operating members interchangeably.

File: sphinx/source/governance.rst (hunk 10 above)
Comment (achamayou) [reply sphinx/source/governance.rst]: Ok

**Comment by cimetrics** [comment]
images
